plutus-ir/src/Language/PlutusIR/Transform/ThunkRecursions.hs

```diff
@@ -0,0 +1,174 @@
+-- compilable to PLC. See Note [Thunking recursions] for details.
+module Language.PlutusIR.Transform.ThunkRecursions (thunkRecursionsTerm) where
+
+import           PlutusPrelude
+
+import           Language.PlutusIR
+import           Language.PlutusIR.MkPir
+import           Language.PlutusIR.Transform.Substitute
+
+import           Language.PlutusCore.Quote
+import qualified Language.PlutusCore.StdLib.Data.Unit   as Unit
+
+import           Control.Lens
+
+import           Data.List
+import qualified Data.Map                               as Map
+import           Data.Maybe
+import           Data.Traversable
+
+{- Note [Thunking recursions]
+Our fixpoint combinators in Plutus Core know how to handle mutually recursive values
+of *function type*. That is, we can define `map : List Int -> List Int` but *not*
+`map : forall a . List a -> List a`, because the latter has a universally
+qualified type, instead of a function type (although it is a function underneath).
+
+We could solve this problem for universally quantified values by lifting the forall
+out of the recursion. This is a bit like performing the following transformation:
+
+    map : forall a . List a -> List a
+    map f xs = case xs of
+        [] -> []
+        x:xs -> f x : map f xs
+
+vs
+
+    -- non-recursive
+    map : forall a . List a -> List a
+    map = map'
+        where
+            -- recursive, but *monomorphic* in the 'a' we instantiate to, so of
+            -- simple function type
+            map' : List a -> List a
+            map' f xs = case xs of
+                [] -> []
+                x:xs -> f x : map' f xs
+
+However, this has two drawbacks:
+- It only works for polymorphic functions. There are other kinds of non-function
+  values which we might want to define recursively.
+- It doesn't work if we use polymorphic recursion, because the function we are
+  defining is monomorphic, so the recursive call must be at the same type.
+
+The latter is quite annoying, because in practice all interesting functions over
+irregular datatypes will use polymorphic recursion, and we've gone to some lengths
+to allow ourselves to define irregular datatypes.
+
+The alternative solution is: given a proposed recursive definition of a value of
+non-function type, we simply *make* it into a value of function type, by adding
+a unit argument and forcing it at all the uses in the body.
+
+So we do something like this:
+
+    -- non-recursive, acts as an "adaptor" for other consumers of the original function
+    map : forall a . List a -> List a
+    map = map' () @a
+        where
+            -- recursive, but thunked, so of simple function type
+            map' : () -> forall b . List b -> List b
+            map' f xs = case xs of
+                [] -> []
+                x:xs -> f x : (map' () @b) f xs
+
+This has the advantage of always working, but it's annoying because we have to go
+in and edit the body of the function definition.
+-}
```

Comment:
Would be nice to see a description of the actual algorithm, because you're doing quite a lot of things below.

Reply:
Hm, okay.

Reply:
But please read what I wrote below first. Maybe it makes sense and then we can use it.

Reply:
Ah, you've just answered below.